backoffcache.go

```diff
@@ -0,0 +1,257 @@
+package discovery
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-peerstore/addr"
+)
+
+// BackoffDiscovery is an implementation of discovery that caches peer data and attenuates repeated queries
+type BackoffDiscovery struct {
+	disc         discovery.Discovery
+	strat        BackoffFactory
+	peerCache    map[string]*backoffCache
+	peerCacheMux sync.RWMutex
+}
+
+func NewBackoffDiscovery(disc discovery.Discovery, strat BackoffFactory) (discovery.Discovery, error) {
+	return &BackoffDiscovery{
+		disc:      disc,
+		strat:     strat,
+		peerCache: make(map[string]*backoffCache),
+	}, nil
+}
+
+type backoffCache struct {
+	nextDiscover time.Time
+	prevPeers    map[peer.ID]peer.AddrInfo
+
+	peers      map[peer.ID]peer.AddrInfo
+	sendingChs map[chan peer.AddrInfo]int
+
+	ongoing bool
+	strat   BackoffStrategy
+	mux     sync.Mutex
+}
+
+func (d *BackoffDiscovery) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	return d.disc.Advertise(ctx, ns, opts...)
+}
+
+func (d *BackoffDiscovery) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
```

Comment:
this is a gigantic method, can we break it up?

Reply:
No problem. Splitting out the two goroutines should decrease the size significantly. Let me know if you think that's still too big, or if you want me to break up the goroutines more.

Reply:
Just did this ^